src/libtls/tls_server.c

```diff
@@ -671,6 +671,12 @@ static status_t process_certificate(private_tls_server_t *this,
 		return NEED_MORE;
 	}
 	certs = bio_reader_create(data);
+	if (!certs->remaining(certs))
+	{
+		DBG1(DBG_TLS, "no cert sent by peer");
```

Comment:
If the client does not have any certificate to send it sends a _Certificate_ message without any certificate in it (certificate length == 0). In TLS 1.3 the connection is terminated by the server when no _CertificateVerify_ message is present.

In legacy TLS `process_cert_verify()`, which calls `tls_find_public_key()`, is never called because a client sends a _ChangeCipherSpec_ message between its _KeyExchange_ and _Finished_ messages. tls_fragmentation.c sets status to `FAILED` in the `process()` method since the `cipherspec_changed()` evaluates FALSE because `this->peer` is not null and the current state is `STATE_KEY_EXCHANGE_RECEIVED` and not `STATE_CERT_VERIFY_RECEIVED`. This means if a TLS 1.2 client tries to open a connection without a client certificate the handshake hangs after server output `processing TLS ChangeCipherSpec record (1 bytes)` which does not look right to me. The above change terminates the connection to the client.

Reply:
As discussed in today's meeting, we leave the current code since it is functional but open a new issue to investigate the case that the connection hangs after a _ChangeCipherSpec_ message.